### PR TITLE
include:irq Modification decided on no-os meeting

### PR DESCRIPTION
**Pull request by mchindri**
I didn't test modifications on xilinx.
Integration of the uart irq will be done in an other pull request.

Later Edit:
The last commit didn't looks good because of how git works. The changes I have done in xilinx/irq.c are:
 - Make static irq_unregister
 - Move irq_unregister before irq_register_callback (will be used inside it)
 - Add this in irq_register_callback:
```
if (!callback_desc)
		return irq_unregister(desc, irq_id);
```

### CODE REVIEW

**Comment by mchindri** [comment]
v2
 - fix codacy

File: drivers/platform/xilinx/irq.c [hunk 1]
```diff
@@ -230,66 +230,69 @@ int32_t irq_disable(struct irq_ctrl_desc *desc, uint32_t irq_id)
...
- * @brief Register a callback to handle the irq events.
+ * @brief Unregisters a generic IRQ handling function.
  * @param desc - The IRQ controller descriptor.
  * @param irq_id - Interrupt identifier.
- * @param callback_desc - Callback descriptor
  * @return SUCCESS in case of success, FAILURE otherwise.
  */
-int32_t irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
-			      struct callback_desc *callback_desc);
+static int32_t irq_unregister(struct irq_ctrl_desc *desc, uint32_t irq_id)
```
Comment (cristiantpop) [inline drivers/platform/xilinx/irq.c]: Why would you change the order of the functions(irq_unregister and irq_unregister)? It is hard to see the changes if you do this.

File: drivers/platform/xilinx/irq.c [hunk 2]
```diff
@@ -230,66 +230,69 @@ int32_t irq_disable(struct irq_ctrl_desc *desc, uint32_t irq_id)
...
  * @return SUCCESS in case of success, FAILURE otherwise.
  */
-int32_t irq_unregister(struct irq_ctrl_desc *desc, uint32_t irq_id)
+int32_t irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
+			      struct callback_desc *callback_desc);
 {
 	struct xil_irq_desc *xil_dev = desc->extra;
 
+	if (!callback_desc)
+		return irq_unregister(desc, irq_id);
```
Comment (cristiantpop) [inline drivers/platform/xilinx/irq.c]: OK, I see, you removed it from the interface

File: drivers/platform/xilinx/irq.c [hunk 3]
```diff
@@ -230,66 +230,69 @@ int32_t irq_disable(struct irq_ctrl_desc *desc, uint32_t irq_id)
...
+
```
Comment (cristiantpop) [inline drivers/platform/xilinx/irq.c]: Maybe, make a pull request with this commit.

File: drivers/platform/xilinx/irq.c (hunk 1 above)
Comment (mchindri) [reply drivers/platform/xilinx/irq.c]: Because irq_unregister is called in irq_resgister_callback. An to avoid writing the prototype.

Comment (dbogdan) [reply drivers/platform/xilinx/irq.c]: irq.c includes irq.h, isn't it?

Comment (mchindri) [reply drivers/platform/xilinx/irq.c]: I will make a new pull request for this

Comment (mchindri) [reply drivers/platform/xilinx/irq.c]: I will make this commit in an other pull request

Comment (mchindri) [reply drivers/platform/xilinx/irq.c]: yes, but the last commit removes the function from there.

Comment (mchindri) [reply drivers/platform/xilinx/irq.c]: I edit the pull request description regarding this changes

**Comment by mchindri** [comment]
todo: rename callbakc_ctx to ctx

**Comment by mchindri** [comment]
v3:
 - Rename callbakc_ctx -> ctx
 - Remove commit that deleted irq_unregister
 - Fix bug found in aducm/irq.c

File: include/irq.h [hunk 4]
```diff
@@ -83,30 +101,21 @@ int32_t irq_ctrl_init(struct irq_ctrl_desc **desc,
...
 
 /* Global interrupt disable */
 int32_t irq_global_disable(struct irq_ctrl_desc *desc);
 
-/* Enable specific interrupt */
-int32_t irq_source_enable(struct irq_ctrl_desc *desc, uint32_t irq_id);
-
-/* Disable specific interrupt */
-int32_t irq_source_disable(struct irq_ctrl_desc *desc, uint32_t irq_id);
-
```
Comment (dbogdan) [inline include/irq.h]: Why did you decide to move those declarations?

Comment (mchindri) [reply include/irq.h]: By mistake, will move them back

**Comment by mchindri** [comment]
v4:
 - Undo move of irq_disable and irq_enable

**Comment by mchindri** [comment]
will merg if no more comments

**Pull request merged**